tests/unit/channel_test.py

```diff
@@ -33,19 +33,24 @@ def ssh_channel(ssh_client_session):
         chan.close()
 
 
-@pytest.mark.xfail(
-    reason='This test causes SEGFAULT, flakily. '
-    'Ref: https://github.com/ansible/pylibssh/issues/57',
-    strict=False,
-)
-@pytest.mark.forked
```

Comment:
@Jakuje did you mean to also delete this `@pytest.mark.forked` marker? It was isolating this test from the rest.

If it is no longer needed indeed, we should also drop `pytest-forked` from all the places: https://github.com/search?q=repo%3Aansible%2Fpylibssh%20forked&type=code.

Reply:
I thought the forked was needed because of the random segfaults. Most of the tests work without that, but indeed not all. The commit cc6bf632332c6f24eef36046967896c80addb17d introducing this line does not show why it was needed. With "forked", its impossible to see the errors from that test, which complicates debugging. Let me try with that again ...

Reply:
Now it crashes only in rpm build for some reason. Is there something specific in the rpm builds that is different from the manylinux build? It fails consistently with different fedora versions, which means different OpenSSH and libssh versions ...

Reply:
@Jakuje the manylinux wheels bundle a statically compiled `libssh` 0.9.6, while RPMs link against the platform-provided `libssh` build. According to https://download.copr.fedorainfracloud.org/results/packit/ansible-pylibssh-658/fedora-39-x86_64/08285674-python-ansible-pylibssh/builder-live.log.gz, F39 has `libssh-devel-0.10.6-2.fc39.x86_64` on Packit.

I see that the RPM tests don't segfault, but the `exec_command()` output is an empty string for some reason:
```python-traceback
+ /usr/bin/python3 -m tox --current-env -q --recreate -e just-pytest -- --deselect tests/unit/scp_test.py::test_get --deselect tests/unit/scp_test.py::test_put
============================= test session starts ==============================
platform linux -- Python 3.12.7, pytest-7.3.2, pluggy-1.2.0
cachedir: .tox/just-pytest/.pytest_cache
rootdir: /builddir/build/BUILD/ansible-pylibssh-1.2.3.dev75+g566ab40
configfile: pytest.ini
testpaths: tests/
plugins: cov-4.0.0, xdist-3.3.1, forked-1.6.0
created: 4/4 workers
4 workers [21 items]

....F................                                                    [100%]
=================================== FAILURES ===================================
______________________________ test_exec_command _______________________________
[gw0] linux -- Python 3.12.7 /usr/bin/python3

ssh_channel = <pylibsshext.channel.Channel object at 0x7f61b14bfc80>

    def test_exec_command(ssh_channel):
        """Test getting the output of a remotely executed command."""
        u_cmd_out = ssh_channel.exec_command('echo -n Hello World').stdout.decode()
        assert u_cmd_out == u'Hello World'  # noqa: WPS302
        # Test that repeated calls to exec_command do not segfault.
    
        # NOTE: Call `exec_command()` once again from another function to
        # NOTE: force it to happen in another place of the call stack,
        # NOTE: making sure that the context is different from one in this
        # NOTE: this test function. The resulting call stack will end up
        # NOTE: being more random.
>       exec_second_command(ssh_channel)

ssh_channel = <pylibsshext.channel.Channel object at 0x7f61b14bfc80>
u_cmd_out  = 'Hello World'

tests/unit/channel_test.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

ssh_channel = <pylibsshext.channel.Channel object at 0x7f61b14bfc80>

    @pytest.mark.forked
    def exec_second_command(ssh_channel):
        """Check the standard output of ``exec_command()`` as a string."""
        u_cmd_out = ssh_channel.exec_command('echo -n Hello Again').stdout.decode()
>       assert u_cmd_out == u'Hello Again'  # noqa: WPS302
E       AssertionError: assert '' == 'Hello Again'
E         - Hello Again

ssh_channel = <pylibsshext.channel.Channel object at 0x7f61b14bfc80>
u_cmd_out  = ''

tests/unit/channel_test.py:40: AssertionError
---------------------------- Captured stdout setup -----------------------------
Generating public/private rsa key pair.
Your identification has been saved in /tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/ssh_client_rsa_key
Your public key has been saved in /tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/ssh_client_rsa_key.pub
The key fingerprint is:
SHA256:cEFCs7vu2RVSC89QLOat1Ug6lxeHFYyqHL4gnbpM24w ansible-pylibssh integration tests key
The key's randomart image is:
+---[RSA 8192]----+
|     .+.o..  =o. |
|       +ooo + o  |
|      oo+=.= o   |
|       +=*B.o    |
|     ..+SB=.     |
|    . +.=. .     |
|    .o.. ..      |
|   o.* o..       |
|    Eo* .        |
+----[SHA256]-----+
Generating public/private rsa key pair.
Your identification has been saved in /tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/ssh_host_rsa_key
Your public key has been saved in /tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/ssh_host_rsa_key.pub
The key fingerprint is:
SHA256:SdrScQOGHGawqyEUKm1dr8bBNtUn7KIbB4jV4ruWiII mockbuild@fd1b674cd54f4b538c6ebdb2012879fa
The key's randomart image is:
+---[RSA 3072]----+
| .  .+=o+o       |
|....o*++ .+ .    |
|o.o+oo= +.oo     |
|o.. o+.O.+..     |
|. . ..*oS.       |
| . o..o..        |
|.... o +         |
|E . + .          |
|.  .             |
+----[SHA256]-----+
---------------------------- Captured stderr setup -----------------------------
debug2: load_server_config: filename /dev/null
debug2: load_server_config: done config len = 1
debug2: parse_server_config_depth: config /dev/null len 1
debug1: sshd version OpenSSH_9.3, OpenSSL 3.1.4 24 Oct 2023
ssh: connect to host 127.0.0.1 port 45793: Connection refused
ssh: connect to host 127.0.0.1 port 45793: Connection refused
ssh: connect to host 127.0.0.1 port 45793: Connection refused
debug1: private host key #0: ssh-rsa SHA256:SdrScQOGHGawqyEUKm1dr8bBNtUn7KIbB4jV4ruWiII
debug1: setgroups() failed: Operation not permitted
debug1: rexec_argv[0]='/usr/sbin/sshd'
debug1: rexec_argv[1]='-D'
debug1: rexec_argv[2]='-f'
debug1: rexec_argv[3]='/dev/null'
debug1: rexec_argv[4]='-o'
debug1: rexec_argv[5]='LogLevel=DEBUG3'
debug1: rexec_argv[6]='-o'
debug1: rexec_argv[7]='HostKey=/tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/ssh_host_rsa_key'
debug1: rexec_argv[8]='-o'
debug1: rexec_argv[9]='PidFile=/tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/sshd.pid'
debug1: rexec_argv[10]='-o'
debug1: rexec_argv[11]='UsePAM=yes'
debug1: rexec_argv[12]='-o'
debug1: rexec_argv[13]='PasswordAuthentication=no'
debug1: rexec_argv[14]='-o'
debug1: rexec_argv[15]='ChallengeResponseAuthentication=no'
debug1: rexec_argv[16]='-o'
debug1: rexec_argv[17]='GSSAPIAuthentication=no'
debug1: rexec_argv[18]='-o'
debug1: rexec_argv[19]='StrictModes=no'
debug1: rexec_argv[20]='-o'
debug1: rexec_argv[21]='PermitEmptyPasswords=yes'
debug1: rexec_argv[22]='-o'
debug1: rexec_argv[23]='PermitRootLogin=yes'
debug1: rexec_argv[24]='-o'
debug1: rexec_argv[25]='Protocol=2'
debug1: rexec_argv[26]='-o'
debug1: rexec_argv[27]='HostbasedAuthentication=no'
debug1: rexec_argv[28]='-o'
debug1: rexec_argv[29]='IgnoreUserKnownHosts=yes'
debug1: rexec_argv[30]='-o'
debug1: rexec_argv[31]='Port=45793'
debug1: rexec_argv[32]='-o'
debug1: rexec_argv[33]='ListenAddress=127.0.0.1'
debug1: rexec_argv[34]='-o'
debug1: rexec_argv[35]='AuthorizedKeysFile=/tmp/pytest-of-mockbuild/pytest-0/popen-gw0/test_exec_command0/sshd/authorized_keys'
debug1: rexec_argv[36]='-o'
debug1: rexec_argv[37]='AcceptEnv=LANG LC_*'
debug1: rexec_argv[38]='-o'
debug1: rexec_argv[39]='Subsystem=sftp internal-sftp'
Warning: Permanently added '[127.0.0.1]:45793' (RSA) to the list of known hosts.
- generated xml file: /builddir/build/BUILD/ansible-pylibssh-1.2.3.dev75+g566ab40/.test-results/pytest/results.xml -
============================= slowest 10 durations =============================
28.25s setup    tests/unit/channel_test.py::test_double_close
24.28s setup    tests/integration/sshd_test.py::test_sshd_addr_fixture_port
22.69s setup    tests/unit/channel_test.py::test_send_signal
20.16s setup    tests/unit/channel_test.py::test_request_exec
17.74s setup    tests/unit/channel_test.py::test_channel_exit_status
13.68s setup    tests/unit/channel_test.py::test_read_bulk_response
13.57s setup    tests/unit/channel_test.py::test_recv_eof
7.16s setup    tests/unit/channel_test.py::test_send_eof
6.38s setup    tests/unit/sftp_test.py::test_make_sftp
5.75s setup    tests/unit/sftp_test.py::test_put
=========================== short test summary info ============================
FAILED tests/unit/channel_test.py::test_exec_command - AssertionError: assert...
======================== 1 failed, 20 passed in 51.25s =========================
```

Interestingly, we have RPM tests both in Packit and GHA. And the GHA ones pass while the Packit ones fail. The GHA build logs reveal the same libssh version being used.

So there must be something additional. The entire GHA pipeline builds an sdist in one job and then, all the following builds happen from that sdist. Packit also builds an sdist first and feeds that to an RPM spec. However, it does not reuse the sdist from GHA so that build is separate. Still, I can't imagine it causing this problem.

Reply:
Thank you for having a look into that! I am wondering if it could be just a timing issue, that the runners in one case just return from `exec_command()` without the callbacks being able to kick in and record the `stdout`. There is the `ssh_channel_get_exit_status()`, which waits with default timeout, but the exit code is not checked in the test. Let see if the exit code is recorded correctly.

Reply:
ok, the exit status looks ok.

Reply:
Should be resolved now.